File: capacity-to-ship.js
// TIme, Space - O(NlgN), O(1)
/**
 * @param {number[]} weights
 * @param {number} D
 * @return {number}
 */

var shipWithinDays = function(weights, D) {
    let low = Number.MIN_VALUE;
    let high = 0;
    
    for(let weight of weights) {
        high += weight;
        low = Math.max(weight, low);
    }
    
    
   
    
    while(low < high) {
        let mid = low + Math.floor((high - low) / 2);
        
        if(isBSValid(mid, weights, D)) {
            high = mid;
        }
        else {
            low = mid + 1;    
        }
    }
    
    return low;
    
};

function isBSValid(mid, weights, D) {

    let days = 0;
    let sum = 0;
    for(let i=0;i<weights.length;i++) {
        sum += weights[i];
        if(sum > mid) {            
            days++;
            sum = weights[i];
            if(days>=D) {
                return false;
            }
        }
    }
    return true;
}

File: capacityToShip.py
#Time: O(logn)
#Space: O(1)
#Program ran on leetcode successfully

class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        
        if not weights:
            return 1
        
        min_cap = max(weights)
        max_cap = sum(weights)
        
        while min_cap < max_cap:
            curr_cap = min_cap + (max_cap - min_cap) // 2
            
            current_load = 0
            num_days = 1
            
            for i in range(len(weights)):
                if weights[i] + current_load > curr_cap:
                    current_load = weights[i]
                    num_days += 1
                else:
                    current_load += weights[i]
                    
            if num_days <= days:
                max_cap = curr_cap
            else:
                min_cap = curr_cap + 1
                
        return min_cap